Jamus contrib1

jamuspsi opened the pull request:
Hi!  I really really loved your mod, and it let me play for the first time in peaceful without feeling like I was missing out.  The ability to industrialize my supply chain was what drew me, but what really did me in was how ADORABLE the chickens are.

So I made more: sand, glass, log, leather, string.  Each of them is textured in the same goofy way.  I'm not exactly an artist, but I'm pretty proud of how they came out.

Having gone through a whole playthrough using chickens, the thing that ended up being most irritating was having to go to the nether so early before I could start putting them to use.  When I tried to figure out where to insert my new ones, a different sort of tree emerged, and so I redid the breeding lines and the naturally spawning.

I made a couple other small changes too.  I dramatically reduced the frequency of chicken sounds (both steps and clucks), which seems to help with lag somewhat when you have LOTS of them.  I also tweaked the drops so they now drop their respective items, EXCEPT the white and string spiders, which drop bones and spider eyes respectively.

Lastly, one developer to another, your code is very clean and nicely organized.  That was REALLY helpful, because this is the first minecraft project I've ever worked on.  And the first java, actually.  Thanks again for the mod, even if you don't like my changes!

setycz commented:
Hi. Thank you for your interest. Your changes look good at the first sight, but I'll check them in more detail. I have to note, that your textures look really amazing. I would like to include your improvements but I finally finished henhouse which, sadly, introduced several conflicts :(. Can you please check it and update sources?

PS: Thank you for your feedback, to be honest this is my second minecraft (and Java, .NET rules;] ) project;].

setycz commented on src/main/java/com/setycz/chickens/chicken/EntityChickensChicken.java:
```diff
@@ -161,4 +167,47 @@ public void readFromNBT(NBTTagCompound tagCompund) {
     }
+    
+    @Override
+    public int getTalkInterval()
```
Nice! I hate their sounds as well ;]

setycz commented on src/main/java/com/setycz/chickens/chicken/EntityChickensChicken.java:
```diff
@@ -161,4 +167,47 @@ public void readFromNBT(NBTTagCompound tagCompund) {
+    	ItemStack dropItemStack = getChickenDescription().createLayItem();
+    	Item dropItem = dropItemStack.getItem();
+    	
+    	if(dropItem == Items.string) {
```
I like the idea behind it. However, I would prefer to move this logic in ChickensRegistryItem class to a method called 'createDropItem' to encapsulate the logic there.

setycz commented on src/main/java/com/setycz/chickens/ChickensRegistryItem.java:
```diff
@@ -98,7 +98,7 @@ public int getDyeMetadata() {
 
     public boolean canSpawn() {
-        return getTier() == 1 && !isDye();
+        return getTier() == 1 && (!isDye() || isDye(15));
```
Don't use magic constants, simple say EnumDyeColor.BLACK.getMetadata() ;].

setycz replied:
It should be: EnumDyeColor.WHITE.getDyeDamage() , sorry.

setycz commented on src/main/java/com/setycz/chickens/ChickensRegistry.java:
```diff
@@ -11,8 +11,9 @@
     private static Random rand = new Random();
 
-    public static void register(ChickensRegistryItem entity) {
+    public static ChickensRegistryItem register(ChickensRegistryItem entity) {
```
Ahhhhh, why didn't I think of it? Big thumbs up. :+1:

jamuspsi commented:
Cool!  Your notes make sense, and I'll take a stab at rebasing tonight now that henhouse is out (cool btw).

I have one other feature that I haven't submitted yet, which I can either add when I update this branch or leave for later.  I added crafting recipes for each of the new 5 base (flint, snowball, sand , wood), which are 8 eggs surrounding their lay item.  These are made as coloredeggs, so they aren't guaranteed to work (like vanilla eggs).  The idea is if you just can't find one of them spawned in your world, you can make them, but you'll need some assistance from vanilla chickens.  I'm playing with these and it felt really good, not too early at all, imo.

setycz commented:
Nice idea with crafting recipes, may I ask you to submit your improvement in another pull request? I would like to include this one first. 

BTW: I like changes you did in this pull request. I'll have to update my picture of the breeding mechanics to check that it's balanced well. I want to keep certain level of difficulty to make mod challenging. I'll probably add support for breeding configuration so everybody will be able to customise it.

jamuspsi commented:
Okay, this is rebased to not have conflicts and seems to work fine.  (I'll start using it in my game and see if anything comes up.)  I also moved the dropItem like you asked, and fixed my constants to use the enums.

I also also added those recipes I mentioned; let me know if you don't like those in here and I'll comment them out, it's just a few lines to make them go away.  The biggest change is that the metadata of colored_egg is storing a ChickensRegistryItem's id, rather than its dyeMetadata, so that colored eggs can spawn any kind of chickens chicken.

Which may make any existing colored eggs in a map bonkers, so that's an upgrade concern.  As long as you throw those in lava and make new ones, it's fine, though.

jamuspsi commented:
Ah, missed your comment.  Let me unwind some stuff and make that happen.

jamuspsi commented:
![breeding try 1](https://cloud.githubusercontent.com/assets/760756/13907043/2f7f0d92-eeb3-11e5-87a2-e75edf6676be.PNG)
Here's my chart of the breeding.  I ran out of arrows in the free tool I'm using, so it's a bit of a mess, but it should give you the idea.

jamuspsi commented:
Okay, this branch no longer has the crafting recipes for the chickens, and the changes related to the colored eggs is removed (On a separate branch).  Everything tested okay so far (spawning, existing eggs).  Let me know if it looks alright.  (You'll probably need to checkout hard, there were like 3 rebases on this branch)

setycz commented:
Thank fo the changes, I really appreciate that. I'll check them and let you know.

setycz commented on src/main/java/com/setycz/chickens/chicken/EntityChickensChicken.java:
```diff
@@ -171,4 +178,40 @@ public void readFromNBT(NBTTagCompound tagCompund) {
+    @Override
+    protected Item getDropItem()
+    {
+        ChickensRegistryItem chicken = getChickenDescription();
```
Nice! :+1:

setycz commented on src/main/java/com/setycz/chickens/ChickensRegistryItem.java:
```diff
@@ -98,7 +113,7 @@ public int getDyeMetadata() {
 
     public boolean canSpawn() {
-        return getTier() == 1 && !isDye();
+        return getTier() == 1 && (!isDye() || isDye(EnumDyeColor.WHITE.getDyeDamage()));
```
I really liked this change, I didn't think about that it's not possible to obtain bones in peaceful.

setycz commented:
Thank you for the graph of breeding. As a developer I can recommend you yEd (https://www.yworks.com/products/yed), it's free without any limitation. I have few thinks to discuss about the breeding changes:

1] I wanted to have naturally spawning chickens used evenly. However, flint and sand chickens are used much more than the others in your propose. Do you think you can try to balance it better? Otherwise, I'll do than and you can change it in the config file I am going to improve.

2] You basically removed chickens from the nether. I also don't like the nether, but I think it would be good to keep there at least one chicken and make it required for higher tier chickens. So the player is forced going there but he will be rewarded e.g with diamonds ;).

3] I would like to avoid breeding two colour chickens together because I wanted to use it for the rest colours, e.g. to have a gray chicken combining white and black.

I would like to include your changes in the next version and I will also add you officially between mod contributors. Tell me what do you think about it.

jamuspsi commented:
1&2.  Yeah, so as I played through it, I realized that there are a LOT of things using sand and flint.  What's more, gunpowder and iron actually have the same parents!

I think it could make sense to re-add a nether component for certain things.  The obvious ones are ont the right side of the diagram: gold, quartz, diamond, emerald.  Maybe put quartz back (as a simple-ish but pretty building material, it makes a good tier 1 anyway) as tier 1, then...  Hold on, let me make a new chart to run by you in a bit.

3.  I thought you might say that!  Let me try to rework these so that they're not dependent on more than one color.  I wanted them to go down the white line because that's basically where all the peaceful mode replacements are, but I think there's better options for their second.  (I was trying to make sure every color got used at least once.)

3.1. I especially agree because apparently in 1.8.9 brown chickens are practically impossible because jungles basically don't exist.  Any thoughts on somehow making a brown chicken, maybe red+green, corresponding to the color mixing?

setycz commented:
I've manually added some of your changes: white chicken is naturally spawning, chickens are less noisy and some refactoring

jamuspsi commented:
Hi sety.  Sorry for disappearing.  I had a family emergency.  I'm reworking the tree right now, I should be done soonish, and then I'll rebase on the new master.

jamuspsi commented:
![chicken breeding v4](https://cloud.githubusercontent.com/assets/760756/14030589/609fba86-f1d6-11e5-9b26-eb47e5bc2bd9.gif)

Take a look at this and see if it helps address your concerns.  It's slightly harder to read, but I think it addresses your big concern, which was that the nether was completely optional.  I made quartz a tier 1 that spawns in nether (again), but can also be crafted, using quartz, though you still have to go mining in the nether for that.

I added brown as a mix of red and green.  The document shows what things can be gotten without going to the nether- all but one of the tier 4 require it, and all of tier 5.  Diamond moved up a tier to 4.  The dead-ends are redstone, string, leather, netherwart, and all of tier 5.

Let me know what you think!

jamuspsi commented:
Rebased and implemented the tree changes.  Made a couple corrections as I did, including water having three parents and leather and netherwart being the same. (oops).

A couple observations about this line I like: it maintains the iron->gold->diamond progression, as each is needed.  Each of the tier 1s are used at least once (except string, which just didn't thematically fit), but no more than 3 times.  And all the good stuff is behind the nether gate, meaning you'll definitely have to go there to get all of the good stuff.

I'm testing it all now, then I'll rebase the crafting change onto it like before.

setycz commented:
Wow, looks great at the first sight. I'll check it and let you know.

setycz commented:
Hey @jamuspsi . I had holidays, I hope you enjoyed your Easter as well;]. One final think, may I ask you to include yEd project file for the graph you've created? I really like it ;] Thank you

jamuspsi commented:
Sure thing, sometime tonight (it's 6pm my time now) when I get home from
work.
On Apr 1, 2016 10:34 AM, "setycz" <notifications@github.com> wrote:

> Hey @jamuspsi <https://github.com/jamuspsi> . I had holidays, I hope you
> enjoyed your Easter as well;]. One final think, may I ask you to include
> yEd project file for the graph you've created? I really like it ;] Thank you
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/setycz/ChickensMod/pull/22#issuecomment-204439350>
>

jamuspsi commented:
Added the graph, and rebased again on master even though it would have merged fine, because I'm a little obsessive about fast-forward git merges.

setycz commented:
Than you. Seems GitHub haven't done fast-forward merge :(. I'll have to figure out how to set up that.
